enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
+
+### Implementation Details/Notes/Constraints
+
+New downstream OLM operator (openshift/ingress-nodefw-operator) will be designed to do the following:-
+- APIs for the new objecs.
+- controller for each of the new resources to handle add/update/delete
+- verification webhook to validate the configuration.  
+- library to construct and load eBPF ingess maps tables based on the provisioned rules.
+- library for statistics collection and exporting to prometheus.
+
+New downstream repo for the kernel hook implementation (openshift/ingress-nodefw-kernel)
+
+- sample `NodeEndPoint` object
+
+```yaml
+kind: NodeEndpoint
+metadata:
+  name: worker-all-interfaces
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+
+---
+
+kind: NodeEndpoint
+metadata:
+  name: worker-eth0
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+  - "eth0"
+
+---
+```
+
+- sample `IngressNodeFirewall` object
+
+```yaml
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall-on-eth0
+  annotations:
+    node-endpoint: worker-eth0
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 10  
+      protocol: tcp
+      ports: [100]
+      action: allow
+    - order: 20
+      protocol: udp
+      ports: [8000, 8080, 9000]
+      action: allow
+  - fromCIDRS:
+       - 3.3.3.3/24
+    rules:
+    - order: 10
+      protocol: icmp
+      type: 3 # Destination unreachable
+      code: 1 # Host unreachable
+      action: allow
+  - fromCIDRS:
+       - 10:10::1/64
+    rules:
+    - order: 10
+      protocol: icmpv6
+      type: 128 # Ping request
+      action: deny
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 10
+      protocol: tcp
+      action: allow
+
+---
+
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall
+  annotations:
+    node-endpoint: worker-all-interfaces
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 30
+      protocol: udp
+      ports: 8000-9000
+      action: deny
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 20
+      protocol: tcp
+      action: deny
+    - order: 30
+      protocol: udp
+      action: deny
+    - order: 40
+      protocol: sctp
+      action: deny
+```
+
+In case of multiple `IngressNodeFirewall` on the same nodes and interfaces if the CIDR matches or overlap the `order` field will specify which rule applied first, lower order will be excuted first, once the rule is excuted and action is taken firewall operation will be completed, all the following rules won't be checked. 
```

Comment:
I'd rather follow iptables model: order all rules according to their priority/order. And if a rule does not match a packet, then go to the next rule, until one rule matches.  That way, the oder/priority is the most important factor and it's also easier to implement, I think.

We could also match based on longest prefix and use the order as a tie breaker, but that gives the user less flexibility.

Question:
What happens if both CIDR / order overlap? Longest prefix match first?
What happens if CIDR / longest-prefix / order overlap? Random rule applies first?